pkg/util/tags_builder.go

```diff
@@ -0,0 +1,55 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.Datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package util
+
+// TagsBuilder allows to build a slice of tags to generate the context while
+// reusing the same internal slice.
+type TagsBuilder struct {
```

Comment:
Nice implementation. 

my initial implementation idea was the following:

```golang
package tags

type Builder interface {
	Append([]string)
	Copy() []string
}

func NewTagsBuilder() Builder {
	return &builderImpl{}
}

type builderImpl struct {
	data [][]string
}

func (b *builderImpl) Append(tags []string) {
	b.data = append(b.data, tags)
}

func (b *builderImpl) Copy() []string {
	var size int
	for i := range b.data {
		size += len(b.data[i])
	}

	if size == 0 {
		return nil
	}

	output := make([]string, 0, size)
	for i := range b.data {
		output = append(output, b.data[i]...)
	}

	return output
}
```

it use `[][]string` to only store a reference of others slice. and build the final slice only once. 

I'm wondering if you have considered this approach? if yes why did you chose the other approach? This question is only to learn from your experience 😃.

Reply:
This is exactly what we want to avoid actually. The goal is to not create slices for each set of tags when parsing dogstatsd packets. The high number of slice uses a lot of heap in the end and make for longer GC runs.

Also we don't want to hold on to the tagger slice in case it's modify (we come back to the read-only idea where the caller guarantees that he won't change it).

The current implementation create 1 slice per context which is reused for every packet. This means we're allocating 1 slice every 10s instead of slice per dogstatsd message.

Reply:
thanks for the explanation.